From db67e0e4184d0e0ad8402f66a2c077ac00b542fa Mon Sep 17 00:00:00 2001
From: thomas lukaszewicz <tluk@chromium.org>
Date: Mon, 19 Jan 2026 01:05:25 -0800
Subject: [PATCH] [Vertical Tabs] Add horizontal animation support to layout manager

This CL adds an enum to TabCollectionAnimatingLayoutManager that
controls whether animate-in and animate-out animations occur in
the vertical or horizontal direction.

This improves the current state of pinned add / remove animations
(see cast below) and will be used to improve the state of
side-by-side animations in follow-up CLs.

See horizontal animations in the pinned collection
http://screencast/cast/NjY5Mjc4MTc0NDQ1NTY4MHxhN2Y4MjUxMC0zMg

Bug: 459824840
Change-Id: Ifd11a8f644eac586ec3aa5dc46d652d63b00ac6c
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7488812
Commit-Queue: Tom Lukaszewicz <tluk@chromium.org>
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1571017}
---

--- a/chrome/browser/ui/views/tabs/vertical/tab_collection_animating_layout_manager.cc
+++ b/chrome/browser/ui/views/tabs/vertical/tab_collection_animating_layout_manager.cc
@@ -31,11 +31,13 @@ DEFINE_OWNED_UI_CLASS_PROPERTY_KEY(gfx::
 
 TabCollectionAnimatingLayoutManager::TabCollectionAnimatingLayoutManager(
     std::unique_ptr<LayoutManagerBase> target_layout_manager,
-    Delegate* delegate)
+    Delegate* delegate,
+    AnimationAxis animation_axis)
     : target_layout_manager_(
           CHECK_DEREF(AddOwnedLayout(std::move(target_layout_manager)))),
       animation_(this),
-      delegate_(delegate) {
+      delegate_(delegate),
+      animation_axis_(animation_axis) {
   // TODO(crbug.com/459824840): Determine the appropriate animation duration.
   // Currently set to match the duration of TabContainerImpl.
   animation_.SetSlideDuration(
@@ -254,7 +256,11 @@ views::ProposedLayout TabCollectionAnima
             value, initial_bounds, target_child.bounds);
       } else {
         gfx::Rect initial_bounds = target_child.bounds;
-        initial_bounds.set_height(0);
+        if (animation_axis_ == AnimationAxis::kVertical) {
+          initial_bounds.set_height(0);
+        } else {
+          initial_bounds.set_width(0);
+        }
         interpolated_child.bounds = gfx::Tween::RectValueBetween(
             value, initial_bounds, target_child.bounds);
       }
@@ -282,7 +288,11 @@ views::ProposedLayout TabCollectionAnima
       // `RemoveNonAnimatingPendingDeleteViews()`.
       views::ChildLayout interpolated_child = start_child;
       gfx::Rect target_bounds = start_child.bounds;
-      target_bounds.set_height(0);
+      if (animation_axis_ == AnimationAxis::kVertical) {
+        target_bounds.set_height(0);
+      } else {
+        target_bounds.set_width(0);
+      }
       interpolated_child.bounds = gfx::Tween::RectValueBetween(
           value, start_child.bounds, target_bounds);
       result.child_layouts.push_back(interpolated_child);
--- a/chrome/browser/ui/views/tabs/vertical/tab_collection_animating_layout_manager.h
+++ b/chrome/browser/ui/views/tabs/vertical/tab_collection_animating_layout_manager.h
@@ -22,6 +22,10 @@
 class TabCollectionAnimatingLayoutManager : public views::LayoutManagerBase,
                                             public gfx::AnimationDelegate {
  public:
+  // Controls along which axis view bounds are animated during animate-in and
+  // animate-out transitions.
+  enum class AnimationAxis { kVertical, kHorizontal };
+
   class Delegate {
    public:
     virtual bool IsViewDragging(const views::View& child_view) const = 0;
@@ -32,7 +36,8 @@ class TabCollectionAnimatingLayoutManage
 
   explicit TabCollectionAnimatingLayoutManager(
       std::unique_ptr<LayoutManagerBase> target_layout_manager,
-      Delegate* delegate = nullptr);
+      Delegate* delegate = nullptr,
+      AnimationAxis animation_axis = AnimationAxis::kVertical);
   TabCollectionAnimatingLayoutManager(
       const TabCollectionAnimatingLayoutManager&) = delete;
   TabCollectionAnimatingLayoutManager& operator=(
@@ -116,6 +121,10 @@ class TabCollectionAnimatingLayoutManage
   double current_offset_ = 1.0;
 
   const raw_ptr<Delegate> delegate_;
+
+  // The axis along which bounds for animate-in and animate-out transitions are
+  // interpolated.
+  AnimationAxis animation_axis_;
 };
 
 #endif  // CHROME_BROWSER_UI_VIEWS_TABS_VERTICAL_TAB_COLLECTION_ANIMATING_LAYOUT_MANAGER_H_
--- a/chrome/browser/ui/views/tabs/vertical/vertical_pinned_tab_container_view.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_pinned_tab_container_view.cc
@@ -26,7 +26,9 @@ VerticalPinnedTabContainerView::Vertical
     : collection_node_(collection_node) {
   auto* layout_manager =
       SetLayoutManager(std::make_unique<TabCollectionAnimatingLayoutManager>(
-          std::make_unique<views::DelegatingLayoutManager>(this)));
+          std::make_unique<views::DelegatingLayoutManager>(this),
+          /*delegate=*/nullptr,
+          TabCollectionAnimatingLayoutManager::AnimationAxis::kHorizontal));
   collection_node->set_remove_child_from_node(base::BindRepeating(
       &TabCollectionAnimatingLayoutManager::AnimateAndDestroyChildView,
       base::Unretained(layout_manager)));
